internal/venice-common/src/main/java/com/linkedin/venice/memory/ClassSizeEstimator.java

```diff
@@ -0,0 +1,245 @@
+    while (parentClass != null) {
+      classHierarchyFromSubclassToParent.add(parentClass);
+      parentClass = parentClass.getSuperclass();
+    }
+
+    // Iterate from the end to the beginning, so we go from parent to sub
+    for (int i = classHierarchyFromSubclassToParent.size() - 1; i >= 0; i--) {
+      int classFieldsOverhead = overheadOfFields(classHierarchyFromSubclassToParent.get(i));
+      if (classFieldsOverhead == 0) {
+        continue;
+      }
+      size += classFieldsOverhead;
+
+      if (JAVA_MAJOR_VERSION < 15) {
+        /**
+         * In older Java versions, the field layout would always order the fields from the parent class to subclass.
+         * Within a class, the fields could be re-ordered to optimize packing the fields within the "alignment shadow",
+         * but not across classes of the hierarchy.
+         *
+         * See: https://shipilev.net/jvm/objects-inside-out/#_superclass_gaps
+         */
+        if (i > 0) {
+          /**
+           * We align for each class, except the last one, since we'll take care of it below. BUT, importantly, at this
+           * stage we align by pointer size, NOT by alignment size.
+           */
+          size = roundUpToNearest(size, POINTER_SIZE);
+        }
+      }
+    }
+
+    // We align once at the end no matter the Java version
+    size = roundUpToNearestAlignment(size);
+
+    return size;
+  }
+
+  /** Based on: https://shipilev.net/jvm/objects-inside-out/#_data_types_and_their_representation */
+  private static int getPrimitiveSize(Class<?> c) {
+    if (c.isPrimitive()) {
+      if (c.equals(boolean.class)) {
+        return 1;
+      } else if (c.equals(byte.class)) {
+        return Byte.BYTES;
+      } else if (c.equals(char.class)) {
+        return Character.BYTES;
+      } else if (c.equals(short.class)) {
+        return Short.BYTES;
+      } else if (c.equals(int.class)) {
+        return Integer.BYTES;
+      } else if (c.equals(float.class)) {
+        return Float.BYTES;
+      } else if (c.equals(long.class)) {
+        return Long.BYTES;
+      } else if (c.equals(double.class)) {
+        return Double.BYTES;
+      }
+
+      // Defensive code
+      throw new IllegalStateException(
+          "Class " + c.getSimpleName()
+              + " is said to be a primitive but does not conform to any known primitive type!");
+    } else {
+      throw new IllegalArgumentException("Class " + c.getSimpleName() + " is not a primitive!");
+    }
+  }
+
+  static int roundUpToNearestAlignment(int size) {
+    return roundUpToNearest(size, ALIGNMENT_SIZE);
+  }
+
+  /** Deal with alignment by rounding up to the nearest boundary. */
+  private static int roundUpToNearest(int size, int intervalSize) {
+    int partialAlignmentWindowUsage = size % intervalSize;
+    int waste = partialAlignmentWindowUsage == 0 ? 0 : intervalSize - partialAlignmentWindowUsage;
+    int finalSize = size + waste;
+    return finalSize;
```

Comment:
That's right, and that is the logic I'm using in `HeapSizeEstimatorTest::roundUpToNearestAlignment`, so they are fully equivalent AFAICT. We could pick either one for the main code. I thought this one was slightly more clear, so I picked it for the main code, but we could do the reverse. LMK if you have a preference.

Reply:
I hesitated on this... I already need to invoke or copy a fair amount of the main code into the test code, which leaves me with a feeling that it is somewhat inadequate. It is very likely that `actual == expected` if both of these have been computed with the same code 😅 ... and I especially want to avoid the case where the main code gets a regression introduced, but it is undetected because the test code just calls the main code for both actual and expected paths...

And so, in this particular case, since as part of experimenting I ended up coding two valid and equivalent versions of the algorithm (at least AFAICT) I thought it may be good to let the test code have its own implementation, different from the main code.

Reply:
Sounds good.